### DISCUSSION TITLE
Gestion des contenus

### DISCUSSION THREAD

**Discussion by m-maillot** [SocialGouv/cdtn-admin#906] [Ideas]
Le métier a remonté le besoin de modifier du contenu mais de ne pas le publier de suite. 

Par exemple :
 * Modification d'une contribution
 * Mise à jour d'une fiche information
 * ...

Dans ces cas là, soit on dépublie le contenu, on le modifie puis au moment venu on va le publier à nouveau (cela demande à ce que le contenu ne soit plus disponible pendant cette période). Soit on modifie mais on ne publie plus sur la prod (cela demande un freeze de la mise à jour des données en prod). Soit on modifie dans un fichier à part et le jour j, on vient reporter les modifications dans l'admin.

Ces solutions ne sont pas idéales.

Avec l'arrivé de l'outil de contrib dans l'admin, une idée émerge (venu de Catherine). On a actuellement 3 grandes  fonctionnalités dans l'administration : Alerting, Edition, Publication. La première est bien découpé et n'interfère pas sur le reste. Les deux suivantes (edition et publication) sont fortement liées. L'idée est de délier ces deux fonctionnalités pour avoir une partie édition et une partie publication.

La première étape consiste à décoreller en base de données, la partie édition et publication. Aujourd'hui, si j'édite une page information, je vais sauvegarder mes modifications directement dans la table `documents` qui gère également la publication. On va donc créer une table information qui va contenir les données d'édition. Lorsque l'édition est terminée (lu et appouvé), on va pouvoir mettre son contenu dans la table `documents`. Cette dernière contiendra alors les contenus à jour et prêt pour la publication. 

Voici un schéma expliquant la logique :
<img width="970" alt="image" src="https://user-images.githubusercontent.com/992514/234870109-1175dfcd-50f3-4e6e-a3d1-70a1c107a5da.png">

Pour les contributions, c'est le même principe. Cela permet d'avoir toujours une version que l'on peut modifier sans bloquer la prod et une version valider en prod. 

On retrouve en même temps l'idée d'avoir des tables consistantes pour chaque contenu sans devoir casser la logique actuelle de mise en prod de l'admin. 
On retrouve aussi l'idée d'avoir la possibilité d'avoir une table de suivi des mises à jour en prod. En effet, à chaque fois qu'un document est mis à jour, on va pouvoir alimenter une table qui indique que ce document a été mis à jour avec le diff.

On aura ainsi ce découpage dans l'admin :

 * Alerting
 * Edition
 * Publication

L'alerting contient : Code du travail, Convention collective, Fiches service-public et Fiches travail-emploi
L'édition : les pages infos, le glossaire, Blocs KALI, Fichers, Fiches SP
Publication : Contenus, Thèmes, Requêtes préqualifiés, Mise à jour

**Comment by m-maillot** [comment]
Afin d'apporter du concret, voici ce qui doit être mis en place pour gérer les pages infos : 

1. Création d'une ou plusieurs tables pour sauvegarder l'`editorial_content`. Les tables représenteront la donnée comme on la traite sur l'admin (date, titre, type de section, liste des sections dans une autre table avec jointure...). On peut s'appuyer sur le travail de @Viczei pour la structure des tables.
2. Alimenter les nouvelles tables à partir du format json contenu dans la table document.
3. Mettre à jour l'UI de l'admin pour ne plus sauvegarder en json dans la table document mais directement dans les nouvelles tables.
4. Ajouter un bouton `Publier` sur l'édition. Cette action va mettre à jour le json dans la table document pour l'`editorial_content`.
5. Ajouter une notion d'état pour savoir si le contenu en cours d'édition est en draft (modifié mais pas encore publié) ou en publié. 

On retrouve du coup le découpage entre l'édition avec des tables relationnelles avec du contenu que l'on peut modifier sans impacter la prod et la publication avec la table document. Ici on n'a pas d'impacte sur la publication, le code actuel continuera de fonctionner comme avant.

On a déjà ce principe présent sur les contributions et les modèles de courrier.

**Comment by m-maillot** [comment]
Pour la nouvelle version des contributions, l'idée est la même car l'on a déjà la table séparée où l'on peut éditer les contributions sans impacter la prod. L'idée est de réaliser l'étape 4 ci dessus qui va s'occuper de venir mettre à jour le json de la contribution dans la table `documents` pour pouvoir les passer en prod.

**Reply by maxgfr** [reply]
Yes !

**Comment by maxgfr** [comment]
Prochaine étape : passer le modèle de courrier 😎
